internal/cache/softreservations.go

```
@@ -0,0 +1,243 @@
+func (s *SoftReservationStore) GetSoftReservation(appID string) (*SoftReservation, bool) {
+	s.storeLock.RLock()
+	defer s.storeLock.RUnlock()
+	appSoftReservation, ok := s.store[appID]
+	if !ok {
+		return &SoftReservation{}, ok
+	}
+	return s.deepCopySoftReservation(appSoftReservation), ok
+}
+
+// GetAllSoftReservationsCopy returns a copy of the internal store. As this indicates, this method does a deep copy
+// which is slow and should only be used for purposes where this is acceptable such as tests.
+func (s *SoftReservationStore) GetAllSoftReservationsCopy() map[string]*SoftReservation {
+	s.storeLock.RLock()
+	defer s.storeLock.RUnlock()
+	storeCopy := make(map[string]*SoftReservation, len(s.store))
+	for appID, sr := range s.store {
+		storeCopy[appID] = s.deepCopySoftReservation(sr)
+	}
+	return storeCopy
+}
+
+// CreateSoftReservationIfNotExists creates an internal empty soft reservation for a particular application.
+// This is a noop if the reservation already exists.
+func (s *SoftReservationStore) CreateSoftReservationIfNotExists(appID string) {
+	s.storeLock.Lock()
+	defer s.storeLock.Unlock()
+	_, ok := s.store[appID]
+	if !ok {
+		r := make(map[string]v1beta1.Reservation)
+		sr := &SoftReservation{
+			Reservations: r,
+			Status:       make(map[string]bool),
+		}
+		s.store[appID] = sr
+	}
+}
+
+// AddReservationForPod adds a reservation for an extra executor pod, attaching the associated node and resources to it.
+// This is a noop if the reservation already exists.
+func (s *SoftReservationStore) AddReservationForPod(ctx context.Context, appID string, podName string, reservation v1beta1.Reservation) {
+	s.storeLock.Lock()
+	defer s.storeLock.Unlock()
+	appSoftReservation, ok := s.store[appID]
+	if !ok {
+		svc1log.FromContext(ctx).Info("Could not put reservation since appID does not exist in reservation store", svc1log.SafeParam("appID", appID))
```

I think this method should return an err when the reservation is not in store, as that should never happen

---

Good call, addressed.